pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2017 The OpenPitrix Authors. All rights reserved.
+// Use of this source code is governed by a Apache license
+// that can be found in the LICENSE file.
+
+package qingcloud
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/yunify/qingcloud-sdk-go/client"
+	"github.com/yunify/qingcloud-sdk-go/config"
+	"github.com/yunify/qingcloud-sdk-go/service"
+	"github.com/yunify/qingcloud-sdk-go/utils"
+
+	"openpitrix.io/openpitrix/pkg/cmd/runtime"
+	"openpitrix.io/openpitrix/pkg/logger"
+)
+
+func init() {
+	runtime.RegisterRuntime(new(QingcloudRuntime))
+}
+
+type QingcloudRuntime struct {
+}
+
+func (p *QingcloudRuntime) InitService() (qingcloudService *service.QingCloudService, err error) {
+	userConf, err := config.NewDefault()
+	if err != nil {
+		return
+	}
+	err = userConf.LoadUserConfig()
+	if err != nil {
+		return
+	}
+	qingcloudService, err = service.Init(userConf)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (p *QingcloudRuntime) InitClusterService() (clusterService *service.ClusterService, err error) {
+	qingcloudService, err := p.InitService()
+	if err != nil {
+		logger.Errorf("Failed to init qingcloud api service: %v", err)
+		return
+	}
+	clusterService, err = qingcloudService.Cluster(qingcloudService.Config.Zone)
+	return
+}
+
+func (p *QingcloudRuntime) InitJobService() (jobService *service.JobService, err error) {
+	qingcloudService, err := p.InitService()
+	if err != nil {
+		logger.Errorf("Failed to init qingcloud api service: %v", err)
+		return
+	}
+	jobService, err = qingcloudService.Job(qingcloudService.Config.Zone)
+	return
+}
+
+func (p *QingcloudRuntime) WaitJobs(jobIds []string, timeout time.Duration, waitInterval time.Duration) error {
+	jobService, err := p.InitJobService()
+	if err != nil {
+		logger.Errorf("Failed to init job service when wait cluster jobs %s: %v", jobIds, err)
+		return err
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(jobIds))
+	done := make(chan error, 1)
+
+	for _, jobId := range jobIds {
+		go func() {
+			defer wg.Done()
+			done <- client.WaitJob(jobService, jobId, timeout, waitInterval)
+
+			waitErr, ok := <-done
+			if !ok {
+				return
+			}
+			if waitErr != nil {
+				logger.Errorf("Failed to wait cluster job [%s]: %v", jobId, waitErr)
+				err = waitErr
```

Comment:
it is unsafe for multi goroutine, `err = xxx` is not atomic operate (not sure).

Reply:
May generate multiple job, any one of these jobs fail will cause whole failed. 
chan 1 means only one process can write final err to the result err.